packages/subproviders/src/subproviders/nonce_tracker.ts

```diff
@@ -0,0 +1,98 @@
+import * as _ from 'lodash';
+
+import EthereumTx = require('ethereumjs-tx');
+import ethUtil = require('ethereumjs-util');
+import providerEngineUtils = require('web3-provider-engine/util/rpc-cache-utils');
+
+import { BlockParamLiteral } from '@0xproject/types';
+
+import { ErrorCallback, JSONRPCPayload, NonceSubproviderErrors, OptionalNextCallback } from '../types';
+
+import { Subprovider } from './subprovider';
+
+// We do not export this since this is not our error, and we do not throw this error
+const NONCE_TOO_LOW_ERROR_MESSAGE = 'Transaction nonce is too low';
+export class NonceTrackerSubprovider extends Subprovider {
```

Comment:
Let's add a comment above this class explaining that this was heavily inspired by the ProviderEngine `NonceSubprovider` but with the added feature that the cached nonce is cleared if it ever becomes invalid.